File: app/backend/src/main/java/es/nullpointers/eventvsmerida/entity/Evento.java
package es.nullpointers.eventvsmerida.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.time.LocalDateTime;;

/**
 * Entidad que representa un evento dentro de la aplicación.
 *
 * <p>
 * Contiene la información principal del evento, incluyendo título,
 * descripción, fechas, localización, coordenadas, imagen, usuario creador
 * y categoría asociada.
 * </p>
 *
 * @author Eva Retamar
 * @author David Muñoz
 * @author Adrián Pérez
 */
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "\"Evento\"")
public class Evento {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @NotNull
    @Column(name = "titulo", nullable = false, length = Integer.MAX_VALUE)
    private String titulo;

    @NotNull
    @Column(name = "descripcion", nullable = false, length = Integer.MAX_VALUE)
    private String descripcion;

    @NotNull
    @Column(name = "fecha_inicio", nullable = false)
    private LocalDateTime fechaInicio;

    @NotNull
    @Column(name = "fecha_fin", nullable = false)
    private LocalDateTime fechaFin;

    @NotNull
    @Column(name = "localizacion", nullable = false, length = Integer.MAX_VALUE)
    private String localizacion;

    @Column(name = "latitud", nullable = false)
    private Double latitud;

    @Column(name = "longitud", nullable = false)
    private Double longitud;

    @NotNull
    @Column(name = "foto", nullable = false, length = Integer.MAX_VALUE)
    private String foto;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.SET_DEFAULT)
    @JoinColumn(name = "id_usuario", nullable = false)
    private Usuario usuario;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.SET_DEFAULT)
    @JoinColumn(name = "id_categoria", nullable = false)
    private Categoria categoria;
}

File: app/frontend/lib/models/evento.dart
/// Modelo de datos que representa los eventos de la app.
///
/// @author: Eva Retamar
/// @author: Adrián Pérez
/// @author: David Muñoz
class Evento {
  final int id;
  final String titulo;
  final String descripcion;
  final DateTime fechaInicio;
  final DateTime fechaFin;
  final String localizacion;
  final double? latitud;
  final double? longitud;
  final String foto;
  final String emailUsuario;
  final String nombreCategoria;

  Evento({
    required this.id,
    required this.titulo,
    required this.descripcion,
    required this.fechaInicio,
    required this.fechaFin,
    required this.localizacion,
    required this.latitud,
    required this.longitud,
    required this.foto,
    required this.emailUsuario,
    required this.nombreCategoria,
  });

  factory Evento.fromJson(Map<String, dynamic> json) {
    return Evento(
      id: json['id'] ?? 0,
      titulo: json['titulo'] ?? '',
      descripcion: json['descripcion'] ?? '',
      fechaInicio: DateTime.parse(json['fechaInicio'].toString()),
      fechaFin: DateTime.parse(json['fechaFin'].toString()),
      localizacion: json['localizacion'] ?? '',
      latitud: json['latitud'] != null ? double.tryParse(json['latitud'].toString()) : null,
      longitud: json['longitud'] != null ? double.tryParse(json['longitud'].toString()) : null,
      foto: json['foto'] ?? '',
      emailUsuario: json['emailUsuario'] ?? '',
      nombreCategoria: json['nombreCategoria'] ?? '',
    );
  }
}